File: cpp/Source/calculation_feature_matching.cpp
// -*- coding: utf-8 -*-
/******************************************************************************
Created on Mon Nov 11 14:51:37 2020

@author: Wei Huajing
@company: KAMERAWERK
@e-mail: wei@kamerawerk.cn

@title: Source-Calculation on feature matching
******************************************************************************/

#include "..\Header\operation_vector.h"

#include "..\Header\calculation_feature_matching.h"
#include "..\Header\configuration_global_variable.h"

//------------------------------------------------------------------------------
/*
Calculate horizontal and vertical difference and determine which image is left or right

Args:
	Mat image_a: Mat object of image from camera 0
	Mat image_b: Mat object of image from camera 1
	display: (bool) whether to imshow

Returns:
	horizontal and vertical difference between left and right images
*/
pair<bool, vector<double>> CalculateDifference(Mat& image_a,
												Mat& image_b,
												const string& match_operator,
												bool display = false) {

	cout << endl << "-- Calculate Difference" << endl;
	cout << "==> operator: " << match_operator << endl;

	//feature matching detector
	Ptr<Feature2D>detector;

	if (match_operator == "ORB") {

		detector = ORB::create(NUM_FEATURE_POINTS);
	}
	if (match_operator == "SIFT") {

		detector = SIFT::create(NUM_FEATURE_POINTS);
	}
	if (match_operator == "SURF") {

		detector = SURF::create(HESSION_THRESHOLD);
	}
	//define vector of key points
	vector<KeyPoint> key_points_a;
	vector<KeyPoint> key_points_b;

	//define descriptor
	Mat descriptor_a;
	Mat descriptor_b;

	//detect and compute descriptor
	detector->detectAndCompute(image_a, Mat(), key_points_a, descriptor_a);
	detector->detectAndCompute(image_b, Mat(), key_points_b, descriptor_b);

	//match with Brute Force method or Flann Based
	BFMatcher matcher;
	//FlannBasedMatcher matcher;

	//vector of DMatch which stores information about some of the points in it 
	vector<DMatch> matches;

	//match left and right images
	matcher.match(descriptor_a, descriptor_b, matches, Mat());

	if (display) {

		//display the matching result of first step
		Mat image_matches;

		drawMatches(image_a,
					key_points_a,
					image_b,
					key_points_b,
					matches,
					image_matches);

		namedWindow("Feature Matching (original)", cv::WINDOW_NORMAL);
		resizeWindow("Feature Matching (original)", 1600, 450);
		imshow("Feature Matching (original)", image_matches);
		waitKey(666);
	}
	//slope of all key points
	vector<double> slope_key_points;

	//threshold of slope
	double slope_threshold = 0.6;

	//search interval
	int n_interval = 200;

	//select a suitable slop from this list
	vector<double> vector_slope_candidate;

	//traverse and collect the candidate slope
	for (double this_slope = -slope_threshold; this_slope < slope_threshold; this_slope += (slope_threshold / n_interval)) {

		vector_slope_candidate.push_back(this_slope);
		//cout << this_slope << endl;
	}
	//collect the final slope
	vector_slope_candidate.push_back(slope_threshold);
	//cout << vector_slope_candidate.size() << endl;

	//vector to collect distance sum for each slope 
	for (int k = 0; k < matches.size(); ++k) {

		//left key point
		double x_key_points_a = key_points_a[matches[k].queryIdx].pt.x;
		double y_key_points_a = key_points_a[matches[k].queryIdx].pt.y;

		//right key point
		double x_key_points_b = key_points_b[matches[k].trainIdx].pt.x;
		double y_key_points_b = key_points_b[matches[k].trainIdx].pt.y;

		//diff of x and y
		double diff_x = x_key_points_a - x_key_points_b;
		double diff_y = y_key_points_a - y_key_points_b;

		//slope
		slope_key_points.push_back(diff_y / (diff_x - image_a.cols));

		//cout << k << " " << matches[k].queryIdx << " " << matches[k].trainIdx << endl;
		//cout << slope_key_points[k] << endl;
	}
	//write slope into txt file
	ofstream out_file;
	out_file.open("../Outcome/slope.txt");

	for (int k = 0; k < slope_key_points.size(); k++) {

		out_file << slope_key_points[k] << endl;
	}
	//list to collect distance sum for each slope
	vector<double>vector_distance_for_each_slope;

	//find the best slope
	for (int k = 0; k < vector_slope_candidate.size(); ++k) {

		double this_distance = 0.0;

		for (int kk = 0; kk < slope_key_points.size(); ++kk) {

			this_distance += abs(slope_key_points[kk] - vector_slope_candidate[k]);
		}
		vector_distance_for_each_slope.push_back(this_distance);
	}
	//minimum of the sum of the distance between candidate and real slope counts
	int  index_min = MinimumIndex(vector_distance_for_each_slope);

	//suitable slope from matching result
	double matched_slope = vector_slope_candidate[index_min];
	cout << "==> matched slope: " << matched_slope << endl;

	//list good matched result
	vector<DMatch> good_matches;

	//traverse all slopes and make classification to define good matches
	for (int k = 0; k < slope_key_points.size(); ++k) {

		//the key points whose slope is near matched slop may be considered
		if (abs(slope_key_points[k] - matched_slope) < slope_threshold / n_interval) {

			good_matches.push_back(matches[k]);
		}
	}
	cout << "==> amount of matches: " << matches.size() << endl;
	cout << "==> amount of good matches: " << good_matches.size() << endl;

	//vector to collect x and y shift
	vector<double> vector_x_diff;
	vector<double> vector_y_diff;

	//coordinates from matchesn
	for (int k = 0; k < good_matches.size(); ++k) {

		//left key point
		double x_key_points_a = key_points_a[good_matches[k].queryIdx].pt.x;
		double y_key_points_a = key_points_a[good_matches[k].queryIdx].pt.y;

		//right key point
		double x_key_points_b = key_points_b[good_matches[k].trainIdx].pt.x;
		double y_key_points_b = key_points_b[good_matches[k].trainIdx].pt.y;

		//diff of x and y
		double diff_x = x_key_points_a - x_key_points_b;
		double diff_y = y_key_points_a - y_key_points_b;

		//diff in x and y direction
		vector_x_diff.push_back(diff_x);
		vector_y_diff.push_back(diff_y);

		//cout << diff_x << endl;
		//cout << diff_y << endl;
	}
	double x_shift = VectorAverage(vector_x_diff);
	double y_shift = VectorAverage(vector_y_diff);

	cout << "==> x shift: " << x_shift << endl;
	cout << "==> y shift: " << y_shift << endl;

	//vector to store x y shift
	vector<double> vector_shift;

	vector_shift.push_back(x_shift);
	vector_shift.push_back(y_shift);

	//output result
	pair<bool, vector<double>> pair_output;

	//init
	pair_output.first = true;
	pair_output.second = vector_shift;

	if (good_matches.size() < THRESHOLD_NUM_GOOD_MATCHES) {
	
		pair_output.first = false;
	}
	if (display) {

		Mat image_good_matches;
		drawMatches(image_a,
					key_points_a,
					image_b,
					key_points_b,
					good_matches,
					image_good_matches);

		namedWindow("Feature Matching (good)", WINDOW_NORMAL);
		resizeWindow("Feature Matching (good)", 1600, 450);
		imshow("Feature Matching (good)", image_good_matches);
		waitKey(666);
	}
	return pair_output;
}
int CheckOutOrder(Mat& image_a, 
				Mat& image_b, 
				const string& match_operator) {

	pair<bool, vector<double>>pair_output = CalculateDifference(image_a, image_b, match_operator);

	//invalid output: insufficent matches
	if (!pair_output.first) {
	
		return -1;
	}
	//horizontal difference
	double x_shift = pair_output.second[0];

	if (x_shift > 0) {

		cout << "--> Conclusion: Image A is left, and image B is right." << endl;

		return 1;
	}
	else {

		cout << "--> Conclusion: Image B is left, and image A is right." << endl;

		return 0;
	}
}
vector<Mat> DualCamerasOrder(Mat& image_a,
							Mat& image_b,
							const string& match_operator) {
		
	cout << endl << "-- Dual Order" << endl;
	cout << "==> operator: " << match_operator << endl;

	bool flag = CheckOutOrder(image_a, image_b, match_operator);
	
	//vector to store 2 images
	vector<Mat> vector_image;

	if (flag) {

		vector_image.push_back(image_a);
		vector_image.push_back(image_b);
	}
	else {

		vector_image.push_back(image_b);
		vector_image.push_back(image_a);
	}

	return vector_image;
}
pair<bool, double> CalculateVerticalDifference(Mat& image_left,
												Mat& image_right,
												const string& match_operator) {
	
	cout << endl << "-- Calculate Vertical Difference" << endl;
	cout << "==> operator: " << match_operator << endl;

	//output result
	pair<bool, double> pair_output;

	//fet result from last step
	pair<bool, vector<double>> pair_input = CalculateDifference(image_left, image_right, match_operator, false);
	
	//horizontal difference
	double y_shift = pair_input.second[1];

	//select mutiple of 4
	double y_shift_final = round(y_shift / ACCURACY_Y_SHIFT) * ACCURACY_Y_SHIFT;

	//init result
	pair_output.first = pair_input.first;
	pair_output.second = y_shift;
	//pair_output.second = y_shift_final;

	if (y_shift > 0) {

		cout << "--> Result: Right image is "
			<< abs(y_shift_final)
			<< " (+"
			<< abs(y_shift)
			<< " in reality) "
			<< "pixels higher than left one"
			<< endl;
	}
	else {

		cout << "--> Result: Left image is "
			<< abs(y_shift_final)
			<< " (-"
			<< abs(y_shift)
			<< " in reality) "
			<< "pixels higher than right one"
			<< endl;
	}
	return pair_output;
}
	

File: python/calculation_feature_matching.py
# -*- coding: utf-8 -*-
"""
Created on Fri Nov 13 20:35:12 2020

@author: Wei Huajing
@company: KAMERAWERK
@e-mail: wei@kamerawerk.cn

@title：Module-Feature Matching

"""

import cv2

import numpy as np
from matplotlib import pyplot as plt

from configuration_font import legend_prop,\
                               text_prop,\
                               label_prop,\
                               title_prop,\
                               annotation_prop
                               
def bgr_rgb(img):
    
    (r, g, b) = cv2.split(img)
    
    return cv2.merge([b, g, r])

def DrawMatch(img1, img2, kp1, kp2, match):
    
    outimage = cv2.drawMatches(img1, kp1, img2, kp2, match, outImg=None)
    plt.imshow(bgr_rgb(outimage))

def DrawMatchedLines(key_points_left,key_points_right,image_left):
    
    for k in range(len(key_points_left)):
        
        plt.plot([key_points_left[k].pt[0],key_points_right[k].pt[0]+np.shape(image_left)[1]],
                 [key_points_left[k].pt[1],key_points_right[k].pt[1]],
                 '-',
                 linewidth=.13)
        
def MergeDualImages(image_left,image_right):
    
    if np.shape(image_left)!=np.shape(image_right):
        
        print('=> ERROR: Incorrect Dimension of Dual Images')
        
        return
    
    image_merged=np.zeros((np.shape(image_left)[0],
                           np.shape(image_left)[1]*2,
                           3))
    
    #left image
    image_merged[:,:np.shape(image_left)[1],:]=image_left[:,:,:]
    image_merged[:,np.shape(image_left)[1]:np.shape(image_left)[1]*2,:]=image_right[:,:,:]
          
    return image_merged.astype(image_left[0,0,0])

def DrawDualImages(image_left,image_right):
    
    plt.imshow(bgr_rgb(MergeDualImages(image_left,image_right)))
    
def KeyPointsFromMatches(key_points_left,
                         key_points_right,
                         matches):
    
    key_points_left_sorted=[]
    key_points_right_sorted=[]
    
    for this_match in matches:

        #obtain key points from matching result
        key_points_left_sorted.append(key_points_left[this_match.queryIdx])
        key_points_right_sorted.append(key_points_right[this_match.trainIdx])
 
        # print(matches.index(this_match),this_match.queryIdx,this_match.trainIdx)
        
    return key_points_left_sorted,key_points_right_sorted

def SlopeFromKeyPoints(key_points_left,key_points_right,image_left):
    
    slope_key_points=[]
    
    for k in range(len(key_points_left)):
        
        point_left=key_points_left[k]
        point_right=key_points_right[k]
        
        diff_x=point_left.pt[0]-point_right.pt[0]
        diff_y=point_left.pt[1]-point_right.pt[1]
    
        #calculate slope of key points
        this_slope=diff_y/(diff_x-np.shape(image_left)[1])

        slope_key_points.append(this_slope)
    
    return slope_key_points 

def DrawMatchedPoints(image_left,
                      key_points_left,
                      key_points_right):

    #draw good match on left image
    for this_left_point in key_points_left:
        
        plt.plot(this_left_point.pt[0],
                 this_left_point.pt[1],
                 'r.',
                 markersize=2.3)
        
    #draw good match on right image
    for this_right_point in key_points_right:
        
        plt.plot(this_right_point.pt[0]+np.shape(image_left)[1],
                 this_right_point.pt[1],
                 'b.',
                 markersize=2.3)
        
def CalculateGoodMatches(slope_key_points,matches):
    
    #max slope
    slope_threshold=0.6
    n_interval=200
    
    #list to collect distance sum for each slope 
    list_distance_for_each_slope=[]
    
    #select a suitable slop from this list
    list_slope_candidate=list(np.linspace(-slope_threshold,slope_threshold,n_interval))
    
    #find the best slope
    for this_slope_candidate in list_slope_candidate:
        
        this_distance=0
        
        for this_slope in slope_key_points:
            
            this_distance+=np.abs(this_slope-this_slope_candidate)
    
        list_distance_for_each_slope.append(this_distance)
        
    #minimum of the sum of the distance between candidate and real slope counts
    index_min=list_distance_for_each_slope.index(np.min(list_distance_for_each_slope))
   
    #suitable slope from matching result
    matched_slope=list_slope_candidate[index_min]
    
    print('==> matched slope:',matched_slope)
    
    plt.figure(figsize=(13,6))
    
    #list good matched result
    good_matches=[]
    
    #traverse all slopes and make classification
    for k in range(len(slope_key_points)):
        
        this_slope=slope_key_points[k]
        
        #the key points whose slope is near matched slop may be considered
        if abs(this_slope-matched_slope)<slope_threshold/n_interval:
    
            good_matches.append(matches[k])
            
            plt.plot(k,this_slope,'k.')
            
        else:
                
            plt.plot(k,this_slope,'g.')

    plt.hlines(y=matched_slope-3*slope_threshold/n_interval,
               xmin=0,
               xmax=len(slope_key_points),
               color='r',
               linestyle='--')
    
    plt.hlines(y=matched_slope+3*slope_threshold/n_interval,
               xmin=0,
               xmax=len(slope_key_points),
               color='r',
               linestyle='--')
    
    plt.vlines(x=0,
               ymin=matched_slope-3*slope_threshold/n_interval,
               ymax=matched_slope+3*slope_threshold/n_interval,
               color='r',
               linestyle='--')
    
    plt.vlines(x=len(slope_key_points),
               ymin=matched_slope-3*slope_threshold/n_interval,
               ymax=matched_slope+3*slope_threshold/n_interval,
               color='r',
               linestyle='--')
    
    plt.ylim([-slope_threshold,slope_threshold])
    
    plt.ylabel('Slope (-)',fontdict=label_prop)
    plt.xlabel('Key Points (-)',fontdict=label_prop)

    plt.title('Slope Distribution of all Key Points',fontdict=title_prop)
    
    #set ticks fonts
    plt.tick_params(labelsize=12)
    labels=plt.gca().get_xticklabels()+plt.gca().get_yticklabels()
    
    #label fonts
    [this_label.set_fontname('Times New Roman') for this_label in labels]
    
    print('==> amount of matches:',len(matches))
    print('==> amount of good matches:',len(good_matches))
    
    return good_matches
    
def CalculateVerticalDifference(key_points_left,key_points_right):
    
    #y coordinate of key points
    y_key_points_left=[this_key_point.pt[1] for this_key_point in key_points_left]
    y_key_points_right=[this_key_point.pt[1] for this_key_point in key_points_right]
    
    return np.average(np.array(y_key_points_left)-np.array(y_key_points_right))
    
def CalculateHorizontalDifference(key_points_left,key_points_right):
    
    #x coordinate of key points
    x_key_points_left=[this_key_point.pt[0] for this_key_point in key_points_left]
    x_key_points_right=[this_key_point.pt[0] for this_key_point in key_points_right]
    
    return np.average(np.array(x_key_points_left)-np.array(x_key_points_right))